### DISCUSSION TITLE
Returning default(T?) shows null warning even though T is nullable

### DISCUSSION THREAD

**Discussion by AustinBryan** [dotnet/csharplang#1611] [Q&A] [answered]
Was just playing around on SharpLab and found this:

```csharp
public static T? Foo<T>() => default;    // Cannot convert null literal to non-nullable reference
public static string? Bar() => default;    // Perfectly fine
```

This seems counter intuitive to me. Maybe I'm not understanding this right. 

Both `default`s are returning `null` and both `return` types are nullable, so neither should give an error, right?

I saying `T?` or `(T)?` By that I mean, if I did `Foo<int>` would `T` be `int?` or `int`? And I did Foo<int?> is `T` `int??` or `int?`? (not like that would make any meaningful sense). 

I would expect that if I passed either `int` or `int?` to `T?`, `T` would be an `int?`. 
Is this a bug or by intent because of the implicit casting that happens from `T?` to `T`?

But if it is the latter, why is it there no warning for `string?`?

Also `Foo<T?>` is an error, so that doesn't solve this.

**Comment by stakx** [comment, marked as answer]
> if I did `Foo<int>` would `T` be `int?` or `int`?

It would be `int`.

>And I did `Foo<int?>` is `T` `int??` or `int?`?

There's no such thing as `int??`. Using `T?` requires you to add a generic parameter type constraint `where T : struct` to your method, because `Nullable<T>` likewise constraints `T` to be `struct` and that constraint is "propagated" to your method. It appears that a `where T : struct` explicitly disallows `T` to be a nullable value type (my guess is to prevent nested or multiply-nullable types), so once you add the constraint, you'll no longer be able to use e.g. `int?` as a generic type argument.

**Comment by Mteheran** [comment]
Try to use T as structure:

public static T? Foo<T>() where T:struct => default(T?);

**Comment by AustinBryan** [comment]
@stakx I know there isn't, that's why I said:

> not like that would make any meaningful sense. 

If I did that though, wouldn't I not be able to use `string?`?

**Comment by Mteheran** [comment]
unfortunately,  "string?" is not available in the current version. It is gonna be included in C# 8
